statediff/service.go

```diff
@@ -0,0 +1,337 @@
+			}
+			// Send a payload to subscribers only if isn't empty
+			if !isEmpty {
+				sds.send(payload)
+			}
+		case err := <-errCh:
+			log.Warn("Error from state change event subscription, breaking loop", "error", err)
+			sds.close()
+			return
+		case <-sds.QuitChan:
+			log.Info("Quitting the statediffing process")
+			sds.close()
+			return
+		}
+	}
+}
+
+// processStateChanges builds the state diff Payload from the modified accounts in the StateChangeEvent
+func processStateChanges(stateChanges state.StateChanges, block *types.Block) (Payload, error) {
+	var accountDiffs []AccountDiff
+	var emptyPayload Payload
+
+	// Iterate over state changes to build AccountDiffs
+	for addr, modifiedAccount := range stateChanges {
+		a, err := buildAccountDiff(addr, modifiedAccount)
+		if err != nil {
+			return emptyPayload, err
+		}
+
+		accountDiffs = append(accountDiffs, a)
+	}
+
+	stateDiff := StateDiff{
+		BlockNumber:     block.Number(),
+		BlockHash:       block.Hash(),
+		UpdatedAccounts: accountDiffs,
+	}
+
+	stateDiffRlp, err := rlp.EncodeToBytes(stateDiff)
+	if err != nil {
+		return emptyPayload, err
+	}
+	payload := Payload{
+		StateDiffRlp: stateDiffRlp,
+	}
+
+	return payload, nil
+}
+
+// buildAccountDiff
+func buildAccountDiff(addr common.Address, modifiedAccount state.ModifiedAccount) (AccountDiff, error) {
+	emptyAccountDiff := AccountDiff{}
+	accountBytes, err := rlp.EncodeToBytes(modifiedAccount.Account)
+	if err != nil {
+		return emptyAccountDiff, err
+	}
+
+	var storageDiffs []StorageDiff
+	for k, v := range modifiedAccount.Storage {
+		// Storage diff value should be an RLP object too
+		encodedValueRlp, err := rlp.EncodeToBytes(v[:])
+		if err != nil {
+			return emptyAccountDiff, err
+		}
+		storageKey := k
+		diff := StorageDiff{
+			Key:   storageKey[:],
+			Value: encodedValueRlp,
+		}
+		storageDiffs = append(storageDiffs, diff)
+	}
+
+	address := addr
+	return AccountDiff{
+		Key:     address[:],
+		Value:   accountBytes,
+		Storage: storageDiffs,
+	}, nil
+}
+
+func (sds *Service) addressInWatchedAddresses(address common.Address) bool {
+	for _, watchedAddress := range sds.WatchedAddresses {
+		checkSummedWatchedAddress := common.HexToAddress(watchedAddress).Hex()
+		if checkSummedWatchedAddress == address.Hex() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (sds *Service) filterByWatchedAddresses(stateChanges state.StateChanges) state.StateChanges {
+	if len(sds.WatchedAddresses) > 0 {
+		filteredStateChanges := make(state.StateChanges)
+		for addr, modifiedAccount := range stateChanges {
+			if sds.addressInWatchedAddresses(addr) {
+				filteredStateChanges[addr] = modifiedAccount
+			}
+		}
+		return filteredStateChanges
+	} else {
+		return stateChanges
+	}
+}
+
+func getEmptyStateDiffRlp(block *types.Block) ([]byte, error) {
+	stateDiffWithoutUpdatedAccounts := StateDiff{
+		BlockNumber: block.Number(),
+		BlockHash:   block.Hash(),
+	}
+
+	return rlp.EncodeToBytes(stateDiffWithoutUpdatedAccounts)
```

Comment:
Probably nbd but I'm wondering whether we might be able to avoid doing this encoding for every block. Do you think we might be able to recognize a payload as empty earlier? Thinking for example that maybe `processStateChanges` could recognize if `StateDiff. UpdatedAccounts` is empty, and either return a specific error/boolean so that we could assign `isEmpty` without needing to call `isEmptyPayload`

Reply:
👍 good eye!